app/hooks/use-countdown-timer.ts

```diff
@@ -0,0 +1,85 @@
+import { useEffect, useState } from "react"
+
+type useCountdownTimerReturnValue = {
+  timeLeft: number
+  startCountdownTimer: (seconds: number, callback?: () => void) => void
+  stopCountdownTimer: () => void
+  resetCountdownTimer: (seconds: number, newCallback?: () => void) => void
+  resetCountdownTimerAndExecuteCallback: (
+    seconds: number,
+    newCallback?: () => void,
+  ) => void
+  stopCountdownTimerAndExecuteCallback: (newCallback?: () => void) => void
+}
+
+export const useCountdownTimer = (): useCountdownTimerReturnValue => {
+  const [timeLeft, setTimeLeft] = useState<number>()
+  const [completedCallback, setCompletedCallback] = useState<() => void>()
+
+  useEffect(() => {
+    if (timeLeft !== undefined) {
+      const interval = setInterval(() => {
+        setTimeLeft(timeLeft - 1)
+      }, 1000)
+
+      if (timeLeft === 0) {
+        clearInterval(interval)
+        if (completedCallback) {
+          completedCallback()
+        }
+      }
+
+      return () => {
+        clearInterval(interval)
+      }
+    }
+    // We need to return from all code paths but we only need a callback if an interval is set so returning undefined is fine
+    return undefined
+  }, [timeLeft, completedCallback])
+
+  const startCountdownTimer = (seconds: number, callback?: () => void) => {
+    setTimeLeft(seconds)
+    if (callback) {
+      setCompletedCallback(() => callback)
+    }
+  }
+
+  const stopCountdownTimer = () => {
```

Comment:
This will trigger the callback which I dont think is the expected behavior. ```stopCountdownTimer``` to me implies that you are pausing the time at its current value.

Reply:
Yup agreed, changed